src/http_server.rs

```diff
@@ -0,0 +1,604 @@
+        *the_error = Some(Box::new(err));
+    }
+}
+
+#[derive(Clone)]
+struct ErrorTrackerArc(Arc<ErrorTracker>);
+
+impl ErrorTrackerArc {
+    fn new() -> ErrorTrackerArc {
+        ErrorTrackerArc(Arc::new(ErrorTracker::new()))
+    }
+    fn has_error(&self) -> bool {
+        self.0.has_error.load(Ordering::Acquire)
+    }
+}
+
+impl Into<Option<io::Error>> for ErrorTrackerArc {
+    fn into(self) -> Option<io::Error> {
+        let the_error = self.0.the_error.lock().unwrap();
+        let the_error = {
+            if let Some(the_error) = &*the_error {
+                the_error
+            } else {
+                return None;
+            }
+        };
+        if let Some(the_error) = the_error.downcast_ref::<io::Error>() {
+            Some(io::Error::new(the_error.kind(), format!("{}", the_error)))
+        } else {
+            Some(io::Error::new(ErrorKind::Other, format!("{}", the_error)))
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+struct EsthriRejection {
+    message: String,
+}
+
+impl warp::reject::Reject for EsthriRejection {}
+
+impl EsthriRejection {
+    fn warp_rejection<MsgT: AsRef<str>>(message: MsgT) -> warp::Rejection {
+        warp::reject::custom(EsthriRejection {
+            message: message.as_ref().to_owned(),
+        })
+    }
+    fn warp_result<T, MsgT: AsRef<str>>(message: MsgT) -> Result<T, warp::Rejection> {
+        Err(EsthriRejection::warp_rejection(message))
+    }
+}
+
+#[derive(Serialize)]
+struct ErrorMessage {
+    code: u16,
+    message: String,
+}
+
+fn with_bucket(bucket: String) -> impl Filter<Extract = (String,), Error = Infallible> + Clone {
+    warp::any().map(move || bucket.clone())
+}
+
+fn with_s3_client(
+    s3_client: S3Client,
+) -> impl Filter<Extract = (S3Client,), Error = Infallible> + Clone {
+    warp::any().map(move || s3_client.clone())
+}
+
+pub fn esthri_filter(
+    s3_client: S3Client,
+    bucket: &str,
+) -> impl Filter<Extract = (http::Response<Body>,), Error = warp::Rejection> + Clone {
+    warp::path::full()
+        .and(with_s3_client(s3_client))
+        .and(with_bucket(bucket.to_owned()))
+        .and(warp::query::<Params>())
+        .and(warp::header::optional::<String>("if-none-match"))
+        .and_then(download)
+}
+
+pub async fn run(
+    s3_client: S3Client,
+    bucket: &str,
+    address: &SocketAddr,
+) -> Result<(), Infallible> {
+    let routes = esthri_filter(s3_client, bucket).recover(handle_rejection);
+    warp::serve(routes).run(*address).await;
+
+    Ok(())
+}
+
+async fn abort_with_error(
+    archive: Option<&mut Builder<PipeWriter>>,
+    error_tracker: ErrorTrackerArc,
+    err: Report,
+) {
+    let err = {
+        if let Some(archive) = archive {
+            if let Err(err) = archive.finish().await {
+                Report::new(err).wrap_err("closing the archive")
+            } else {
+                err
+            }
+        } else {
+            err
+        }
+    };
```

Comment:
I think this is equivalent 
```suggestion
    let err = match archive.map(|a| a.finish()) {
        Some(Err(err)) => Report::new(err).wrap_err("closing the archive"),
        _ => err,
    };
```